src/shared/http/request.rs

```diff
@@ -0,0 +1,295 @@
+use std::{any::Any, convert::TryFrom, fmt};
+
+use super::{
+  header::{HeaderMap, HeaderName, HeaderValue},
+  method::Method,
+  version::Version,
+  Extensions, Uri,
+};
+use crate::Result;
+
+pub struct Request {
+  head: Parts,
+  body: Vec<u8>,
+}
+
+/// Component parts of an HTTP `Request`
+///
+/// The HTTP request head consists of a method, uri, version, and a set of
+/// header fields.
+#[non_exhaustive]
+pub struct Parts {
+  /// The request's method
+  pub method: Method,
+
+  /// The request's URI
+  pub uri: Uri,
+
+  /// The request's version
+  pub version: Version,
+
+  /// The request's headers
+  pub headers: HeaderMap<HeaderValue>,
+
+  /// The request's extensions
+  pub extensions: Extensions,
+}
+
+/// An HTTP request builder
+///
+/// This type can be used to construct an instance or `Request`
+/// through a builder-like pattern.
+#[derive(Debug)]
+pub struct Builder {
+  inner: Result<Parts>,
+}
+
+impl Request {
+  /// Creates a new blank `Request` with the body
+  #[inline]
+  pub fn new(body: Vec<u8>) -> Request {
+    Request {
+      head: Parts::new(),
+      body,
+    }
+  }
+
+  /// Returns a reference to the associated HTTP method.
+  #[inline]
+  pub fn method(&self) -> &Method {
+    &self.head.method
+  }
+
+  /// Returns a reference to the associated URI.
+  #[inline]
+  pub fn uri(&self) -> &Uri {
+    &self.head.uri
+  }
+
+  /// Returns the associated version.
+  #[inline]
+  pub fn version(&self) -> Version {
+    self.head.version
+  }
+
+  /// Returns a reference to the associated header field map.
+  #[inline]
+  pub fn headers(&self) -> &HeaderMap<HeaderValue> {
+    &self.head.headers
+  }
+
+  /// Returns a reference to the associated extensions.
+  #[inline]
+  pub fn extensions(&self) -> &Extensions {
+    &self.head.extensions
+  }
+  /// Returns a reference to the associated HTTP body.
+  #[inline]
+  pub fn body(&self) -> &Vec<u8> {
+    &self.body
+  }
+
+  /// Consumes the request returning the head and body parts.
+  #[inline]
+  pub fn into_parts(self) -> (Parts, Vec<u8>) {
```

Comment:
`Parts` is private for users. Do you think it could be public?

Reply:
`Parts` in `response` seems to be private either. I suppose they should re-export as `ResponseParts` and `RequestParts`